File: frontend/src/types/crm.ts
export interface Customer {
  id: string;
  name: string;
  customer_type: 'COMPANY' | 'INDIVIDUAL';
  gst_number?: string;
  industry?: string;
  segment: 'ENTERPRISE' | 'SME' | 'VIP' | 'DISTRIBUTOR' | 'GOVERNMENT';
  credit_limit: number;
  payment_terms: string;
  territory_id?: string;
  status: 'ACTIVE' | 'INACTIVE';
  created_at: string;
  contacts?: CustomerContact[];
}

export interface CustomerContact {
  id: string;
  customer_id: string;
  first_name: string;
  last_name: string;
  email?: string;
  phone?: string;
  job_title?: string;
  is_primary: boolean;
}

export interface CustomerAddress {
  id: string;
  customer_id: string;
  address_type: 'BILLING' | 'SHIPPING';
  address_line1: string;
  address_line2?: string;
  city: string;
  state: string;
  country: string;
  zip_code: string;
}

export interface CustomerNote {
  id: string;
  customer_id: string;
  note: string;
  created_by_id: string;
  created_at: string;
}

export interface CustomerActivityLog {
  id: string;
  customer_id: string;
  activity_type: 'CALL' | 'EMAIL' | 'MEETING' | 'OPPORTUNITY_CHANGED';
  description: string;
  created_by_id: string;
  created_at: string;
}

export interface Lead {
  id: string;
  first_name: string;
  last_name: string;
  company_name?: string;
  email?: string;
  phone?: string;
  source: 'WEBSITE' | 'REFERRAL' | 'CAMPAIGN' | 'MANUAL';
  status: 'NEW' | 'CONTACTED' | 'QUALIFIED' | 'UNQUALIFIED' | 'CONVERTED';
  score: number;
  assigned_to_id?: string;
  converted_customer_id?: string;
  created_at: string;
}

export interface LeadActivity {
  id: string;
  lead_id: string;
  activity_type: 'CALL' | 'EMAIL' | 'MEETING' | 'NOTE';
  details: string;
  created_by_id: string;
  created_at: string;
}

export interface OpportunityProduct {
  id: string;
  opportunity_id: string;
  product_id: string;
  quantity: number;
  unit_price: number;
  is_upsell: boolean;
}

export interface Opportunity {
  id: string;
  name: string;
  customer_id: string;
  stage: 'PROSPECTING' | 'QUALIFICATION' | 'PROPOSAL' | 'NEGOTIATION' | 'WON' | 'LOST';
  probability: number;
  expected_revenue: number;
  close_date: string;
  risk_level: 'LOW' | 'MEDIUM' | 'HIGH';
  competitors?: string;
  assigned_to_id?: string;
  lost_reason?: string;
  created_at: string;
  products?: OpportunityProduct[];
}

export interface PricingRule {
  id: string;
  name: string;
  customer_segment: string;
  product_id?: string;
  discount_percentage: number;
  start_date: string;
  end_date: string;
  priority: number;
  status: 'ACTIVE' | 'INACTIVE';
}

export interface SalesTerritory {
  id: string;
  name: string;
  region: string;
  assigned_salesperson_id?: string;
}

export interface QuotationItem {
  id: string;
  quotation_id: string;
  product_id: string;
  quantity: number;
  unit_price: number;
  tax_rate: number;
  total_cost: number;
}

export interface Quotation {
  id: string;
  quotation_number: string;
  opportunity_id?: string;
  customer_id: string;
  status: 'DRAFT' | 'SUBMITTED' | 'APPROVED' | 'REJECTED' | 'ACCEPTED' | 'CONVERTED';
  version: number;
  valid_until: string;
  subtotal: number;
  tax_amount: number;
  discount_amount: number;
  total_amount: number;
  created_at: string;
  items?: QuotationItem[];
}

export interface SalesOrderItem {
  id: string;
  sales_order_id: string;
  product_id: string;
  variant_id?: string;
  quantity: number;
  unit_price: number;
  tax_rate: number;
  total_cost: number;
}

export interface SalesOrder {
  id: string;
  order_number: string;
  customer_id: string;
  quotation_id?: string;
  status: 'DRAFT' | 'PENDING_APPROVAL' | 'APPROVED' | 'SHIPPED' | 'COMPLETED' | 'CANCELLED';
  subtotal: number;
  tax_amount: number;
  discount_amount: number;
  total_amount: number;
  created_by_id: string;
  salesperson_id?: string;
  shipping_status: 'PENDING' | 'SHIPPED' | 'DELIVERED';
  created_at: string;
}

export interface CustomerDocument {
  id: string;
  customer_id: string;
  file_name: string;
  file_path: string;
  file_size: number;
  uploaded_by_id: string;
}

export interface CRMTask {
  id: string;
  title: string;
  description?: string;
  due_date: string;
  priority: 'LOW' | 'MEDIUM' | 'HIGH';
  status: 'PENDING' | 'IN_PROGRESS' | 'COMPLETED';
  assigned_to_id?: string;
  customer_id?: string;
  lead_id?: string;
  created_at: string;
}

export interface CRMMeeting {
  id: string;
  title: string;
  description?: string;
  start_time: string;
  end_time: string;
  location?: string;
  assigned_to_id?: string;
  customer_id?: string;
  lead_id?: string;
}

File: backend/app/schemas/crm.py
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field

class CustomerContactBase(BaseModel):
    first_name: str = Field(..., max_length=100)
    last_name: str = Field(..., max_length=100)
    email: Optional[str] = Field(None, max_length=255)
    phone: Optional[str] = Field(None, max_length=50)
    job_title: Optional[str] = Field(None, max_length=100)
    is_primary: bool = False

class CustomerContactCreate(CustomerContactBase):
    pass

class CustomerContactResponse(CustomerContactBase):
    id: UUID
    customer_id: UUID
    model_config = ConfigDict(from_attributes=True)

class CustomerBase(BaseModel):
    name: str = Field(..., max_length=255)
    customer_type: str = "COMPANY"
    gst_number: Optional[str] = Field(None, max_length=50)
    industry: Optional[str] = Field(None, max_length=100)
    segment: str = "SME"
    credit_limit: float = 0.0
    payment_terms: str = "NET30"
    territory_id: Optional[UUID] = None
    status: str = "ACTIVE"

class CustomerCreate(CustomerBase):
    contacts: Optional[List[CustomerContactCreate]] = None

class CustomerResponse(CustomerBase):
    id: UUID
    created_at: datetime
    contacts: List[CustomerContactResponse] = []
    model_config = ConfigDict(from_attributes=True)

class LeadBase(BaseModel):
    first_name: str = Field(..., max_length=100)
    last_name: str = Field(..., max_length=100)
    company_name: Optional[str] = Field(None, max_length=255)
    email: Optional[str] = Field(None, max_length=255)
    phone: Optional[str] = Field(None, max_length=50)
    source: str = "MANUAL"
    status: str = "NEW"
    score: float = 0.0
    assigned_to_id: Optional[UUID] = None

class LeadCreate(LeadBase):
    pass

class LeadResponse(LeadBase):
    id: UUID
    converted_customer_id: Optional[UUID] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class OpportunityProductBase(BaseModel):
    product_id: UUID
    quantity: float = 1.0
    unit_price: float

class OpportunityProductCreate(OpportunityProductBase):
    pass

class OpportunityProductResponse(OpportunityProductBase):
    id: UUID
    is_upsell: bool
    model_config = ConfigDict(from_attributes=True)

class OpportunityBase(BaseModel):
    name: str = Field(..., max_length=255)
    customer_id: UUID
    stage: str = "PROSPECTING"
    probability: float = 10.0
    expected_revenue: float
    close_date: date
    risk_level: str = "LOW"
    competitors: Optional[str] = Field(None, max_length=255)
    assigned_to_id: Optional[UUID] = None

class OpportunityCreate(OpportunityBase):
    products: Optional[List[OpportunityProductCreate]] = None

class OpportunityResponse(OpportunityBase):
    id: UUID
    lost_reason: Optional[str] = None
    created_at: datetime
    products: List[OpportunityProductResponse] = []
    model_config = ConfigDict(from_attributes=True)

class QuotationItemCreate(BaseModel):
    product_id: UUID
    quantity: float
    unit_price: float

class QuotationItemResponse(BaseModel):
    id: UUID
    product_id: UUID
    quantity: float
    unit_price: float
    tax_rate: float
    total_cost: float
    model_config = ConfigDict(from_attributes=True)

class QuotationCreate(BaseModel):
    customer_id: UUID
    opportunity_id: Optional[UUID] = None
    items: List[QuotationItemCreate]
    valid_until: date
    discount_amount: float = 0.0

class QuotationResponse(BaseModel):
    id: UUID
    quotation_number: str
    customer_id: UUID
    opportunity_id: Optional[UUID] = None
    status: str
    version: int
    valid_until: date
    subtotal: float
    tax_amount: float
    discount_amount: float
    total_amount: float
    created_at: datetime
    items: List[QuotationItemResponse] = []
    model_config = ConfigDict(from_attributes=True)

class SalesOrderItemCreate(BaseModel):
    product_id: UUID
    variant_id: Optional[UUID] = None
    quantity: float
    unit_price: float

class SalesOrderCreate(BaseModel):
    customer_id: UUID
    quotation_id: Optional[UUID] = None
    items: List[SalesOrderItemCreate]
    discount_amount: float = 0.0

class SalesOrderResponse(BaseModel):
    id: UUID
    order_number: str
    customer_id: UUID
    quotation_id: Optional[UUID] = None
    status: str
    subtotal: float
    tax_amount: float
    discount_amount: float
    total_amount: float
    shipping_status: str
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)

class CRMTaskCreate(BaseModel):
    title: str = Field(..., max_length=255)
    description: Optional[str] = None
    due_date: datetime
    priority: str = "LOW"
    customer_id: Optional[UUID] = None
    lead_id: Optional[UUID] = None

class CRMTaskResponse(BaseModel):
    id: UUID
    title: str
    description: Optional[str]
    due_date: datetime
    priority: str
    status: str
    assigned_to_id: Optional[UUID]
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)
